feat(tests): run integration tests with PX4 internal SIH simulation

MaEtUgR opened the pull request:
### Solved Problem
CI integration tests failing intermittently which are hard to reproduce locally and don't directly represent real-world use.

### Solution
Work in progress ⚙️
- Use SIH instead of gazebo classic for basic integration tests. We need to remove the setup and runtime complexity of running Gazebo for normal integration tests.
- Go through issues with the tests themselves. What I found so far:
   - The basic test compared local coordinate z values against relative altitude which is the altitude above home.
   - The basic test has timing issues where it captures the altitude at an arbitrary time when it just goes over a threshold and then keeps checking against that which is highly timing dependent.
- Improve SIH
  - Certain noise amount is not realistic e.g. the accelerometer noise is so high it makes velocity estimation and control really inaccurate.
  - I found that the home position is jumping up and down in SIH because of some combined problem with

### Changelog Entry
```
Run integration tests with PX4 internal SIH simulation
```

### Test coverage
The way I'm currently testing this is with this command:
```
make px4_sitl && DONT_RUN=1 make px4_sitl mavsdk_tests && test/mavsdk_tests/mavsdk_test_runner.py --speed-factor 10 --abort-early --model iris test/mavsdk_tests/configs/sitl.json --verbose --force-color --case 'Takeoff and hold position' --iterations 1
```
Nicer integration for simpler use still to come.

### Context
Similar to #24237 and #25978
My goal is to make this a smaller step not changing too much at once because otherwise it will never be finished.

MaEtUgR committed:
test(mavsdk): switch tests to SIH simulator

mrpollo commented on test/mavsdk_tests/autopilot_tester.cpp:
```diff
@@ -225,13 +225,13 @@ void AutopilotTester::wait_until_hovering()
 void AutopilotTester::wait_until_altitude(float rel_altitude_m, std::chrono::seconds timeout, float delta)
 {
-	auto prom = std::promise<void> {};
+	auto prom = std::promise<void>{};
```
looks like this is breaking a few builds in CI

julianoes committed:
feat(simulation): implement required failure hooks

julianoes committed:
test(mavsdk): leave tests backwards compatible but enable SIH

This reverts some of the changes to keep runners of classic Gazebo in
there but leave it optional, so we can still use it if required.

julianoes committed:
feature(CI): use SIH sim

julianoes committed:
test(ros): add new model_prefix

julianoes committed:
build(cmake): add missing include

julianoes committed:
feat(simulator_sih): implement lockstep by waiting for outputs

Without this, we see test failures in CI as soon as CPU load is higher.

mrpollo committed:
fix(mavsdk_tests): use simulation time for timeouts

Replace 5 functions that used std::future::wait_for() (wall-clock time)
with poll_condition_with_timeout() (simulation time). This prevents
timeouts from expiring prematurely when running at high speed factors
under lockstep.

Affected functions:
- wait_until_altitude()
- wait_for_landed_state()
- wait_for_flight_mode()
- start_and_wait_for_mission_sequence()
- wait_until_speed_lower_than()

Also remove debug printf statements from start_checking_altitude().

mrpollo committed:
fix(mavsdk_tests): use simulation time for sleeps in basics test

Replace std::this_thread::sleep_for() with tester.sleep_for() to use
simulation-aware sleep that accounts for speed factor under lockstep.

julianoes committed:
fix(rcS): scale COM_DISARM_PRFLT by sim speed factor

Symmetric with the existing COM_DL_LOSS_T / COM_RC_LOSS_T /
COM_OF_LOSS_T / COM_OBC_LOSS_T scalings just above. At high simulation
speed the default 10s preflight disarm is reached much sooner in
real wall time than tests expect, so scale it the same way.

julianoes committed:
fix(lockstep_scheduler): fix signal-loss race and ABBA deadlock

Two related thread-safety fixes for the cond_timedwait + set_absolute_time
dance, both surfaced under TSan with multi-instance Mavlink and a fast
sim clock advance:

1. Signal loss between releasing _timed_waits_mutex and entering
   pthread_cond_wait. set_absolute_time could have already broadcast
   to a waiter that hadn't actually started waiting yet, and the
   broadcast would be missed -> wait blocks forever.

   Fix: cond_timedwait uses pthread_cond_timedwait with a short
   wall-clock timeout (10 ms) and re-checks the timeout flag. Lost
   signals turn into a maximum of one loop iteration of latency.

2. ABBA between (passed_lock -> _timed_waits_mutex) used by
   cond_timedwait to mark `done`, and (passed_lock under
   _timed_waits_mutex) used by set_absolute_time to broadcast. TSan
   flagged the inversion immediately.

   Fix: split set_absolute_time into three phases. Phase 1 marks
   timed_outs and stages waiters onto a per-call signal_next list,
   under _timed_waits_mutex only. Phase 2 broadcasts to each waiter,
   outside _timed_waits_mutex but under a new _signaling_mutex held
   for the duration. Phase 3 clears _setting_time under the
   _timed_waits_mutex again.

   The waiter's "dance" (when it sees _setting_time still true on
   exit) acquires _signaling_mutex first, then _timed_waits_mutex,
   guaranteeing it cannot return — and let its stack-local
   passed_lock/passed_cond go out of scope — until set_absolute_time
   has finished signaling.

The TimedWait::timeout flag also becomes std::atomic<bool> since it
is now read by cond_timedwait without holding _timed_waits_mutex.

julianoes committed:
refactor(mavsdk_tests): use sim time directly for execute_mission timeout

poll_condition_with_timeout polls the autopilot's monotonic time, which is
already simulation time. The previous code divided a sim-time budget by the
speed factor to compute a wall-clock timeout, but this is unnecessary and
adds a dependency on the speed-factor query.

julianoes committed:
fix(mavsdk_tests): bump MAVSDK version

julianoes committed:
fix(build): move mavsdk_tests target out of gazebo-classic cmake file

The ExternalProject_Add(mavsdk_tests ...) lived at the bottom of
sitl_targets_gazebo-classic.cmake, outside the if(gazebo_FOUND) gate
and with a comment noting it doesn't depend on sitl_gazebo. Historical
artifact: mavsdk_tests is a black-box MAVLink test driver that talks
to a running PX4 over UDP, and works against any simulator backend
(sitl_gazebo-classic, gz, sih, jmavsim, ...).

With CI moving from classic Gazebo to SIH, the misplacement becomes
actively misleading. Pull the block into a new top-level
cmake/mavsdk_tests.cmake and include it from CMakeLists.txt under the
existing posix-only block, so the dependency graph reflects what is
actually true.

Behavior is unchanged: target is EXCLUDE_FROM_ALL, only built on
demand via `make mavsdk_tests` or `make tests_integration`.

julianoes commented:
I did a bunch of work here over the last few weeks. I need to review the commits one by one myself as well.

hamishwillee commented on docs/uk/test_and_ci/integration_testing_mavsdk.md:
```diff
@@ -45,10 +45,10 @@ The tests need the MAVSDK C++ library installed system-wide (e.g. in `/usr/lib`
 ## Запуск усіх PX4 тестів
 
-To run all SITL tests as defined in [sitl.json](https://github.com/PX4/PX4-Autopilot/blob/main/test/mavsdk_tests/configs/sitl.json), do:
+To run all SITL tests using the SIH simulator as defined in [sih-sitl.json](https://github.com/PX4/PX4-Autopilot/blob/main/test/mavsdk_tests/configs/sih-sitl.json), do:
```
Remove the uk/ko/zh path changes - only en is needed. The rest round trip through crowdin

julianoes replied:
Ok, removed.

julianoes committed:
fix(rcS): use awk instead of bc for scaled timeout math

bc isn't installed in the px4-dev CI container we now use. awk is
required by POSIX so it's always present. Same float math, no extra
dependency.

julianoes commented:
@Claudio-Chies I've taken your mutex fix and force pushed again. I think you're rebase was based on outdated main and had conflicts. You force pushed it over my state which had already fixed the conflicts which was probably not what you intended.

My goal here was to:
1. Get the tests passing in CI.
2. Extract fixes in small group and review and merge them in other PRs until this one becomes easy to review and absorb.

julianoes committed:
fix(mavlink): accept spec-defined params in mission item validation

The param validation table rejected mission items that QGC-generated
plans legitimately contain:

- NAV_VTOL_LAND param3 is the approach altitude per the MAVLink spec.
  QGC writes it in every VTOL landing pattern; PX4's mission logic
  ignores it, but that is no reason to reject the upload.
- DO_LAND_START param5-7 are the optional lat/lon/alt marker position
  per the spec, which QGC also writes.

Both made any QGC VTOL landing plan fail to upload with
MAV_MISSION_INVALID_PARAMx, surfaced by the SIH SITL VTOL tests.

Signed-off-by: Julian Oes <julian@oes.ch>

julianoes committed:
test(mavsdk): clear unused params in land-start VTOL plan

The DO_LAND_START item carried loiter params (heading required, radius,
and a stray 1 in param3) hand-copied from the adjacent LOITER_TO_ALT
item, and LOITER_TO_ALT itself had param3 set, which the spec defines
as empty. PX4's mission item param validation now rejects such items.

Signed-off-by: Julian Oes <julian@oes.ch>

julianoes commented:
Rebased and running tests locally once more. I need to review the commits though and check what is actually required.

julianoes committed:
test(mavsdk): retry offboard start to beat MAVSDK setpoint watchdog

MAVSDK's OffboardImpl::process_heartbeat() resets the setpoint state to
NotActive when a heartbeat without offboard mode arrives more than 3 s
after _last_started. _last_started is only ever written by start(), so
before the first start() it is epoch zero and the guard is always true:
any heartbeat that slips in between set_*() and start() makes start()
fail with NoSetpointSet.

At speed factor 30 PX4's 1 Hz (sim time) heartbeat arrives every ~33 ms
of wall time, which made 'Offboard attitude control' fail roughly one
run in three. Re-send the setpoint and retry start() in a small loop at
both offboard start sites. Failed 1/3 before, 16/16 after.

This is arguably a MAVSDK bug (_last_started should be initialized when
setpoint streaming begins); a retry keeps the tests robust either way.

Signed-off-by: Julian Oes <julian@oes.ch>

julianoes committed:
test(mavsdk): loosen takeoff altitude reach tolerance to 0.1m

The 0.01m (1cm) tolerance was not reachable within the 15s sim-time
window for the standard_vtol, causing the takeoff-and-hold test to time
out. Use 0.1m, which still verifies altitude hold but is physically
achievable, and stays below the 0.15m hold-monitoring tolerance.

Signed-off-by: Julian Oes <julian@oes.ch>

julianoes committed:
ci(sitl_tests): tolerate gcov negative counts in coverage capture

Switching the SITL tests to the SIH simulator exercises sih.cpp, whose
gcov data contains a spurious negative counter (line 510) that newer
lcov treats as a fatal error, failing the coverage step. Pass
--ignore-errors negative so lcov zeroes the bad count and continues, as
the tool itself suggests.

Signed-off-by: Julian Oes <julian@oes.ch>

julianoes committed:
test(mavsdk): raise RTL disarm timeout to 150s to match siblings

The four older VTOL RTL tests waited only 120s for disarm while running
the same climb-to-550m-and-land missions as the other RTL tests, which
already use 150s. 'RTL direct Mission Land' occasionally tipped just
over 120s of sim time (observed 123.76s), failing wait_until_disarmed.
Align all of them to 150s for margin.

Signed-off-by: Julian Oes <julian@oes.ch>

julianoes committed:
test(mavsdk): retry arming to tolerate transient is_armable flicker

wait_until_ready() polls MAVSDK's is_armable, but a known hysteresis gap
(and estimator heading-reference flicker at startup) means the flag can
read armable while PX4 then denies the arm command with 'no heading
reference'. This made arm() fail intermittently. Retry the arm command
until the autopilot accepts it, so the transient is absorbed while real
failures still time out.

Signed-off-by: Julian Oes <julian@oes.ch>

mrpollo commented:
Green builds finally?

julianoes commented:
Yep, and just passed 100/100 times locally.

julianoes commented:
Just force pushing to fixup the conflict.

mrpollo commented:
Maybe see if you can turn this PR into a stack https://github.github.com/gh-stack/

julianoes commented:
For next time. For now, I would have said this is good to go in.

dakejahl reviewed (approved):
I'm excited to see this come in. @julianoes do we need to update the PR description?

mrpollo commented on .github/workflows/sitl_tests.yml:
```diff
@@ -26,17 +26,14 @@ jobs:
     runs-on: [runs-on,runner=8cpu-linux-x64,image=ubuntu24-full-x64,"run-id=${{ github.run_id }}",spot=false,extras=s3-cache]
     container:
-      image: px4io/px4-dev-simulation-focal:2021-09-08
+      image: px4io/px4-dev:v1.16.0-rc1-258-g0369abd556
```
```suggestion
      image: px4io/px4-dev:v1.17.0
```
Please update to the latest one, https://github.com/PX4/PX4-Autopilot/pkgs/container/px4-dev/862838678?tag=v1.17.0

mrpollo commented on .github/workflows/sitl_tests.yml:
```diff
@@ -55,13 +52,19 @@ jobs:
 
       - name: Download MAVSDK
-        run: wget "https://github.com/mavlink/MAVSDK/releases/download/v$(cat test/mavsdk_tests/MAVSDK_VERSION)/libmavsdk-dev_$(cat test/mavsdk_tests/MAVSDK_VERSION)_ubuntu20.04_amd64.deb"
+        run: wget "https://github.com/mavlink/MAVSDK/releases/download/v$(cat test/mavsdk_tests/MAVSDK_VERSION)/libmavsdk-dev_$(cat test/mavsdk_tests/MAVSDK_VERSION)_ubuntu22.04_amd64.deb"
```
we need to turn the mavsdk download/install into a composable action

julianoes replied:
How/what?

mrpollo commented on .github/workflows/sitl_tests.yml:
```diff
@@ -55,13 +52,19 @@ jobs:
+        shell: bash
+        run: make px4_sitl_default
+
+      - name: Cache Stats after PX4 Firmware
```
verify cache

mrpollo reviewed (approved):
I'm going to leave a few comments only for myself to fix later i will merge as is, and work on some updates to the workflow file to bring it up to date

mrpollo commented:
🚀 thanks for the effort @julianoes @MaEtUgR and everyone else that collaborated

Claudio-Chies commented:
Thanks @julianoes for finishing it!

MaEtUgR commented:
Thanks for @julianoes for doing the by far biggest bulk of the work! 🙏🙏🙏
